Manage installation and configuration of Hashicorp products

blarghmatey opened the pull request:
We use a variety of Hashicorp products and are migrating toward using more of them. This provides a component for managing the common and specific methods of setting up these products including the beginnings of configuration models for managing settings of these products.

blarghmatey committed:
Adding installation of arbitrary products from Hashicorp

- Adds logic to install any number of products from Hashicorp based on a config file listing a mapping of name and version for each product.
- Adds helper functions for mapping CPU architectures and linux distribution names
- Adds facts for determining the CPU architecture based on the LInux distribution family

blarghmatey committed:
Added initial pass of config models for Hashicorp products

blarghmatey committed:
Renaming module space to be more concise

Removed the `ol` prefix to make it more useful outside of the bounds of Open
Learning. Changed the module name to be "Bilder" which is the german word for "images", as well as being phonetically similar to "builder". This is because the code in this project is used to build servers, as well as having Packer configurations for creating server images.

blarghmatey committed:
Added management of Systemd services and updated Vault config model

blarghmatey committed:
Implemented functioning sample deploy for a set of Hashicorp products

blarghmatey committed:
Updating requirements pins

blarghmatey committed:
Removing unused Consul models file after splitting to multiple modules

shaidar commented on src/bilder/components/hashicorp/templates/consul.service.j2:
```diff
@@ -0,0 +1,20 @@
+ExecReload=/bin/kill --signal HUP $MAINPID
+KillSignal=SIGTERM
+KillMode=process
+User=consul
```
Shouldn't this be based on `product.name`?

blarghmatey replied:
Yes, but if a product.name doesn't == "consul" then it means that consul won't work so that's why I went ahead and hard-coded it in the systemd units.

shaidar replied:
In that case should we not hardcode the name of the user being created?

blarghmatey replied:
We kind of do. What might make sense is to make the `name` attribute a protected attribute so that it is clear that it shouldn't be changed during instantiation. I'll make that change.

blarghmatey committed:
Moving name attribute to be private to avoid overwriting

Moved the `name` attribute on `HashicorpProduct` models to be a protected attribute so that it is clear that it shouldn't be changed.

shaidar commented:
Looks good as far as I can tell without spending too much time on all the details. Still not very clear on the `FlexibleBaseModel` and its intended use and feel that it all can use a good amount of Docstrings and overall documentation.
